### ISSUE TITLE
NPE at org.eclipse.jetty.websocket.common.WebSocketSession.open()

### ISSUE THREAD

**Issue by cowwoc** [eclipse/jetty.project#2537]
Version 9.4.10.v20180503

I got this error in the wild. I'm not sure how to reproduce it.

```
java.lang.NullPointerException: null
	at org.eclipse.jetty.websocket.common.WebSocketSession.open(WebSocketSession.java:611)
	at org.eclipse.jetty.websocket.common.WebSocketSession.onOpened(WebSocketSession.java:544)
	at org.eclipse.jetty.io.AbstractConnection.onOpen(AbstractConnection.java:202)
	at org.eclipse.jetty.websocket.common.io.AbstractWebSocketConnection.onOpen(AbstractWebSocketConnection.java:446)
	at org.eclipse.jetty.io.AbstractEndPoint.upgrade(AbstractEndPoint.java:440)
	at org.eclipse.jetty.websocket.client.WebSocketUpgradeRequest.upgrade(WebSocketUpgradeRequest.java:624)
	at org.eclipse.jetty.client.http.HttpChannelOverHTTP.exchangeTerminating(HttpChannelOverHTTP.java:118)
	at org.eclipse.jetty.client.HttpReceiver.terminateResponse(HttpReceiver.java:456)
	at org.eclipse.jetty.client.HttpReceiver.responseSuccess(HttpReceiver.java:410)
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.messageComplete(HttpReceiverOverHTTP.java:316)
	at org.eclipse.jetty.http.HttpParser.handleHeaderContentMessage(HttpParser.java:620)
	at org.eclipse.jetty.http.HttpParser.parseFields(HttpParser.java:1156)
	at org.eclipse.jetty.http.HttpParser.parseNext(HttpParser.java:1427)
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.parse(HttpReceiverOverHTTP.java:172)
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.process(HttpReceiverOverHTTP.java:135)
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.receive(HttpReceiverOverHTTP.java:73)
	at org.eclipse.jetty.client.http.HttpChannelOverHTTP.receive(HttpChannelOverHTTP.java:133)
	at org.eclipse.jetty.client.http.HttpConnectionOverHTTP.onFillable(HttpConnectionOverHTTP.java:155)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:281)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:102)
	at org.eclipse.jetty.io.ssl.SslConnection.onFillable(SslConnection.java:291)
	at org.eclipse.jetty.io.ssl.SslConnection$3.succeeded(SslConnection.java:151)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:102)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:118)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:333)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:310)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:168)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:126)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:366)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:760)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:678)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by joakime** [comment]
Link to specific lines mentioned in stacktrace.

https://github.com/eclipse/jetty.project/blob/c134792d2d8bcf71eaa0b2f413bd10bef1492d2c/jetty-websocket/websocket-common/src/main/java/org/eclipse/jetty/websocket/common/WebSocketSession.java#L610-L611

Sounds like the remoteEndpointFactory is null, but how? *boggle*

**Comment by cowwoc** [comment]
I can only see this happening in one of two ways:

1. Someone invokes `open()` before `doStart()`.
2. Someone extends `WebSocketSession` and overrides `doStart()` in a way that allows `remoteEndpointFactory` to be null

Of course there is a third way (someone uses reflection) but I don't think that is likely here.

**Comment by joakime** [comment]
Unable to replicate this one.

**Comment by cowwoc** [comment]
Okay, so let's just close this issue. If I run into this exception again I can open a new issue.

**Comment by joakime** [comment]
Fair enough. 
If we get more information, just comment here, I'll happily reopen it.

**Comment by cowwoc** [comment]
@joakime So I've gotten this error multiple times now and the only way I can see this happening is threading.

I noticed that you're setting `remoteEndpointFactory` to a non-null value in `doStart()` and but it's null in `start()`. You seem to be assuming that the same thread that invokes `doStart()` will invoke `start()` but in actuality when I get the `null` value it is from the following stacktrace:

```
open:611, WebSocketSession (org.eclipse.jetty.websocket.common)
onOpened:544, WebSocketSession (org.eclipse.jetty.websocket.common)
onOpen:202, AbstractConnection (org.eclipse.jetty.io)
onOpen:446, AbstractWebSocketConnection (org.eclipse.jetty.websocket.common.io)
upgrade:440, AbstractEndPoint (org.eclipse.jetty.io)
upgrade:624, WebSocketUpgradeRequest (org.eclipse.jetty.websocket.client)
exchangeTerminating:118, HttpChannelOverHTTP (org.eclipse.jetty.client.http)
terminateResponse:462, HttpReceiver (org.eclipse.jetty.client)
responseSuccess:416, HttpReceiver (org.eclipse.jetty.client)
messageComplete:316, HttpReceiverOverHTTP (org.eclipse.jetty.client.http)
handleHeaderContentMessage:597, HttpParser (org.eclipse.jetty.http)
parseFields:1219, HttpParser (org.eclipse.jetty.http)
parseNext:1508, HttpParser (org.eclipse.jetty.http)
parse:172, HttpReceiverOverHTTP (org.eclipse.jetty.client.http)
process:135, HttpReceiverOverHTTP (org.eclipse.jetty.client.http)
receive:73, HttpReceiverOverHTTP (org.eclipse.jetty.client.http)
receive:133, HttpChannelOverHTTP (org.eclipse.jetty.client.http)
onFillable:155, HttpConnectionOverHTTP (org.eclipse.jetty.client.http)
succeeded:281, AbstractConnection$ReadCallback (org.eclipse.jetty.io)
fillable:102, FillInterest (org.eclipse.jetty.io)
onFillable:291, SslConnection (org.eclipse.jetty.io.ssl)
succeeded:151, SslConnection$3 (org.eclipse.jetty.io.ssl)
fillable:102, FillInterest (org.eclipse.jetty.io)
run:118, ChannelEndPoint$2 (org.eclipse.jetty.io)
runTask:333, EatWhatYouKill (org.eclipse.jetty.util.thread.strategy)
doProduce:310, EatWhatYouKill (org.eclipse.jetty.util.thread.strategy)
tryProduce:168, EatWhatYouKill (org.eclipse.jetty.util.thread.strategy)
run:126, EatWhatYouKill (org.eclipse.jetty.util.thread.strategy)
run:366, ReservedThreadExecutor$ReservedThread (org.eclipse.jetty.util.thread)
runJob:762, QueuedThreadPool (org.eclipse.jetty.util.thread)
run:680, QueuedThreadPool$2 (org.eclipse.jetty.util.thread)
run:844, Thread (java.lang)
```

In this particular scenario (`WebSocketUpgradeRequest`) I am guessing that the two methods are invoked from different threads. As such, there is no guarantee that the second thread that invokes `start()` will see the non-null value set by the first thread. Meaning, you need to add some sort of thread-synchronization mechanism. Does this make sense?

**Comment by cowwoc** [comment]
@joakime Please reopen this issue.

**Comment by joakime** [comment]
There's been some work around WebSocketClient and bean lifecycle in Jetty 9.4.12.v20180830 please give that a try.

**Comment by cowwoc** [comment]
> Can I see your `WebSocketClient` initialization code?

```
WebSocketClient result = new WebSocketClient(getHttpClient());

result.setMaxIdleTimeout(0);
result.setMaxTextMessageBufferSize(65535 * 2);

result.start();
```

where `getHttpClient()` is defined as:

```
SslContextFactory sslContextFactory = new SslContextFactory();
HttpClient result = new HttpClient(sslContextFactory);
result.setRequestBufferSize(8192);
result.setMaxConnectionsPerDestination(512);
result.start();
```

> As for your stacktrace, it has nothing to do with the start anyway, that's a process that (should have) occurred much earlier then the method call WebSocketUpgradeRequest.upgrade().

The point I was trying to make was that you have one thread that invokes `HttpClient.start()` which sets `remoteEndpointFactory` but then you have a completely different thread which invokes the code seen in my stacktrace which reads `remoteEndpointFactory`. Per my understanding, when multiple threads read/write a non-final, non-volatile variable (`remoteEndpointFactory`) you must have some sort of synchronization mechanism.

> Also, if you are creating your own UpgradeRequest objects for the WebSocketClient.connect() method [...]

I am not.

**Comment by joakime** [comment]
> ```
> WebSocketClient result = new WebSocketClient(getHttpClient());
> 
> result.setMaxIdleTimeout(0);
> result.setMaxTextMessageBufferSize(65535 * 2);
> 
> result.start();
> ```
> 
> where `getHttpClient()` is defined as:
> 
> ```
> SslContextFactory sslContextFactory = new SslContextFactory();
> HttpClient result = new HttpClient(sslContextFactory);
> result.setRequestBufferSize(8192);
> result.setMaxConnectionsPerDestination(512);
> result.start();
> ```

Yup, that's pretty straightforward.
One minor unrelated thing to this issue (and a recent [commit](ce6bc23ccb770e8ae5dedc6ed431eb893f837a61)), make sure you declare your EndpointIdentification algorithm to `HTTPS` on the `SslContextFactory`.

**Comment by cowwoc** [comment]
> make sure you declare your EndpointIdentification algorithm to `HTTPS` on the `SslContextFactory`.

Will do. Thank you!

So, does my point about `remoteEndpointFactory` needing thread synchronization make any sense? :)

**Comment by cowwoc** [comment]
@joakime 

I changed my application code and by total coincidence ended up triggering this error 100% of the time. This allowed me to dig further with debugger breakpoints and here is what I saw:

* A thread created a new `WebSocketSession` using this stack-trace:
```
<init>:173, WebSocketSession (org.eclipse.jetty.websocket.common)
createSession:49, WebSocketSessionFactory (org.eclipse.jetty.websocket.common)
upgrade:578, WebSocketUpgradeRequest (org.eclipse.jetty.websocket.client)
exchangeTerminating:118, HttpChannelOverHTTP (org.eclipse.jetty.client.http)
terminateResponse:462, HttpReceiver (org.eclipse.jetty.client)
responseSuccess:416, HttpReceiver (org.eclipse.jetty.client)
messageComplete:316, HttpReceiverOverHTTP (org.eclipse.jetty.client.http)
handleHeaderContentMessage:591, HttpParser (org.eclipse.jetty.http)
parseFields:1213, HttpParser (org.eclipse.jetty.http)
parseNext:1502, HttpParser (org.eclipse.jetty.http)
parse:172, HttpReceiverOverHTTP (org.eclipse.jetty.client.http)
process:135, HttpReceiverOverHTTP (org.eclipse.jetty.client.http)
receive:73, HttpReceiverOverHTTP (org.eclipse.jetty.client.http)
receive:133, HttpChannelOverHTTP (org.eclipse.jetty.client.http)
onFillable:155, HttpConnectionOverHTTP (org.eclipse.jetty.client.http)
succeeded:305, AbstractConnection$ReadCallback (org.eclipse.jetty.io)
fillable:103, FillInterest (org.eclipse.jetty.io)
onFillable:411, SslConnection$DecryptedEndPoint (org.eclipse.jetty.io.ssl)
onFillable:305, SslConnection (org.eclipse.jetty.io.ssl)
succeeded:159, SslConnection$2 (org.eclipse.jetty.io.ssl)
fillable:103, FillInterest (org.eclipse.jetty.io)
run:118, ChannelEndPoint$2 (org.eclipse.jetty.io)
runTask:333, EatWhatYouKill (org.eclipse.jetty.util.thread.strategy)
doProduce:310, EatWhatYouKill (org.eclipse.jetty.util.thread.strategy)
tryProduce:168, EatWhatYouKill (org.eclipse.jetty.util.thread.strategy)
run:126, EatWhatYouKill (org.eclipse.jetty.util.thread.strategy)
run:366, ReservedThreadExecutor$ReservedThread (org.eclipse.jetty.util.thread)
runJob:765, QueuedThreadPool (org.eclipse.jetty.util.thread)
run:683, QueuedThreadPool$2 (org.eclipse.jetty.util.thread)
run:844, Thread (java.lang)```

* Some time later, **the same thread** invoked:

```
open:608, WebSocketSession (org.eclipse.jetty.websocket.common)
onOpened:544, WebSocketSession (org.eclipse.jetty.websocket.common)
onOpened:209, AbstractConnection (org.eclipse.jetty.io)
onOpen:202, AbstractConnection (org.eclipse.jetty.io)
onOpen:446, AbstractWebSocketConnection (org.eclipse.jetty.websocket.common.io)
upgrade:440, AbstractEndPoint (org.eclipse.jetty.io)
upgrade:624, WebSocketUpgradeRequest (org.eclipse.jetty.websocket.client)
exchangeTerminating:118, HttpChannelOverHTTP (org.eclipse.jetty.client.http)
terminateResponse:462, HttpReceiver (org.eclipse.jetty.client)
responseSuccess:416, HttpReceiver (org.eclipse.jetty.client)
messageComplete:316, HttpReceiverOverHTTP (org.eclipse.jetty.client.http)
handleHeaderContentMessage:591, HttpParser (org.eclipse.jetty.http)
parseFields:1213, HttpParser (org.eclipse.jetty.http)
parseNext:1502, HttpParser (org.eclipse.jetty.http)
parse:172, HttpReceiverOverHTTP (org.eclipse.jetty.client.http)
process:135, HttpReceiverOverHTTP (org.eclipse.jetty.client.http)
receive:73, HttpReceiverOverHTTP (org.eclipse.jetty.client.http)
receive:133, HttpChannelOverHTTP (org.eclipse.jetty.client.http)
onFillable:155, HttpConnectionOverHTTP (org.eclipse.jetty.client.http)
succeeded:305, AbstractConnection$ReadCallback (org.eclipse.jetty.io)
fillable:103, FillInterest (org.eclipse.jetty.io)
onFillable:411, SslConnection$DecryptedEndPoint (org.eclipse.jetty.io.ssl)
onFillable:305, SslConnection (org.eclipse.jetty.io.ssl)
succeeded:159, SslConnection$2 (org.eclipse.jetty.io.ssl)
fillable:103, FillInterest (org.eclipse.jetty.io)
run:118, ChannelEndPoint$2 (org.eclipse.jetty.io)
runTask:333, EatWhatYouKill (org.eclipse.jetty.util.thread.strategy)
doProduce:310, EatWhatYouKill (org.eclipse.jetty.util.thread.strategy)
tryProduce:168, EatWhatYouKill (org.eclipse.jetty.util.thread.strategy)
run:126, EatWhatYouKill (org.eclipse.jetty.util.thread.strategy)
run:366, ReservedThreadExecutor$ReservedThread (org.eclipse.jetty.util.thread)
runJob:765, QueuedThreadPool (org.eclipse.jetty.util.thread)
run:683, QueuedThreadPool$2 (org.eclipse.jetty.util.thread)
run:844, Thread (java.lang)```

* At no point was `WebSocketSession.doStart()` invoked.

I still don't have concrete repro steps but here is some verbose logs from Jetty: 
[jetty.log](https://github.com/eclipse/jetty.project/files/2386152/jetty.log)

I stripped out my application logs, minus a single line that reads "Graceful shutdown initiated..."

I still think some race condition is at play because I only started getting this issue on demand once my application started terminating very quickly on startup. I believe this behavior is somehow triggering the bug in Jetty.

Does this help you reproduce the problem and/or figure out what is going on?

**Comment by cowwoc** [comment]
Sorry, I keep on running into this exception. Did you guys read my last comment? Does it give you enough information to fix this bug or do you need more?

**Comment by cowwoc** [comment]
> You are closing your application and you have a pending websocket connect at the same time.
Not much can be done about that kind of scenario.

Do you mean that I am invoking `WebSocketClient.stop()` while there is a pending websocket connect pending? Or did you mean something else?

I have logic in my shutdown routine that waits for all pending `HttpClient` requests to shut down before invoking `HttpClient.stop()` but I don't have the same protection for websockets.

* What event do I have to wait on before it is safe to invoke `WebSocketClient.stop()`?
* Can't Jetty detect this scenario and throw a more meaningful error message?

**Comment by joakime** [comment]
> Do you mean that I am invoking `WebSocketClient.stop()` while there is a pending websocket connect pending? Or did you mean something else?

Yes, that is kind of what I mean.

There is a nuance of this specific scenario occurring during `AbstractWebSocketConnection.onOpen` which only happens during initial upgrade of a new websocket connection.
Not yet an active connection.

> I have logic in my shutdown routine that waits for all pending `HttpClient` requests to shut down before invoking `HttpClient.stop()` but I don't have the same protection for websockets.

A WebSocket upgraded request is no longer an HTTP request, so HttpClient is not tracking it.
A WebSocket is a long term connection, there is no "wait for pending" logic that can be safely applied to it in a generic way. (like can be applied to HTTP)
Only your application can determine when its safe to close (even forcibly) the connection.

> * What event do I have to wait on before it is safe to invoke `WebSocketClient.stop()`?

Your local WebSocket endpoint(s) would need tell you when their `.onClose()` (or equivalent method) is called.

> * Can't Jetty detect this scenario and throw a more meaningful error message?

There's hundreds (if not thousands) of such places that this would need to be applied where the concurrency between the application and the component lifecycle conflict.

Some advice.

* The WebSocketClient uses a HttpClient.
* If you provide a HttpClient to the WebSocketClient constructor then you must stop the WebSocketClient manually first (before stopping the HttpClient)
* If you just modify the default HttpClient built into the WebSocketClient then just stopping the WebSocketClient is sufficient (as its own internal HttpClient is a child bean of the WebSocketClient in the Jetty component hierarchy)
* Don't stop any component that HttpClient or WebSocketClient needs/uses out from underneath them (eg: Executors, SslContextFactory, Schedulers, etc)

**Comment by joakime** [comment]
@cowwoc is there any need to keep this issue open?

**Comment by cowwoc** [comment]
I'm still not a fan of the difficulty involved ensuring that websockets are closed in a thread-safe manner, but I guess that's as good as we can get for now.

**Comment by gregw** [comment]
@lachlan-roberts Can you test this in the refactored jetty-10 websocket.   It would be good if a websocket client stop() does not generate a NPE even if there are connections outstanding.    It is fine for them to produce some kind of exception, but NPE is a bit brutal.

**Comment by joakime** [comment]
Tasks to implement this ...

* Prevent new connections from being accepted (shutdown ServerConnector).
* Wait ____ (configured time) for any pending Upgrades to complete. (reminder: the connection could exist, and be sitting in a queue to be processed).
* Wait for full CLOSED of remote endpoint.
* Wait for full CLOSED of local endpoint.
* Wait for all dispatched threads to local Endpoint object to finish. (eg: could be using JSR356 and streaming APIs)
* Wait for all references to local Endpoint object to be dereferenced (suitable for GC collection)

Then, and only then, can you consider the WebSocket components ready to be shutdown/stopped.

**Comment by cowwoc** [comment]
As complicated as this sounds from your perspective, I hope you can imagine how impossibly more complicated it sounds from an end-user's perspective. We are playing Russian Roulette every time we invoke `close()`.

I don't have any answers, short of maybe breaking this down into smaller sub-problems and tackling them one at a time. Sorry.

**Comment by joakime** [comment]
Starting in Jetty 9.4.15 (due out soon), you can attach 2 different listeners to the container to help with this.

org.eclipse.jetty.io.Connection.Listener - this gives you open/close for the low level tcp jetty connection that websocket lives on top of
org.eclipse.jetty.websocket.common.WebSocketSessionListener - this gives you open/close for the websocket session layer

These can be used to ensure a mechanism to monitor / hook into the connection level details of the websocket.
